Add UBSAN testing

Similar to the existing address sanitization and thread sanitization it's worth adding undefined behavior sanitization testing:
 https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

Running VVL tests with UBSAN does show failing tests.

Here is an example:
```
[ RUN      ] NegativeCommand.MultiDraw
/home/juaramos/projects/Vulkan-ValidationLayers/layers/core_checks/cc_drawdispatch.cpp:284:96: runtime error: member access within misaligned address 0x7ffdc5bf777d for type 'const struct VkMultiDrawIndexedInfoEXT', which requires 4 byte alignment
0x7ffdc5bf777d: note: pointer points here
 00 00 00 00 00 00 00  01 00 00 00 00 00 00 00  00 00 00 00 01 00 00 00  00 00 00 00 00 00 00 00  00
             ^ 
Command "/home/juaramos/projects/Vulkan-ValidationLayers/build-ci/install/bin/vk_layer_validation_tests" failed with return code -6
```

NOTE: This doesn't need to be a new CI pass. This can just be added to ASAN testing.

GCC / Clang accept the following: `-fsanitize=address,undefined`.

NOTE: I had to set the environment variable `UBSAN_OPTIONS` to `abort_on_error=1:halt_on_error=1` to get lldb to quickly point me to the exact problem when running the tests in vscode.

EDIT: For further reference glslang implemented it in this PR https://github.com/KhronosGroup/glslang/pull/3654

Did a quick attempt in a branch (https://github.com/spencer-lunarg/Vulkan-ValidationLayers/actions/runs/13181233782/job/36792380702)

Getting strange linking errors

`/usr/bin/ld: source/opt/libSPIRV-Tools-opt.a(pass_manager.cpp.o):(.data.rel+0x58): undefined reference to `typeinfo for spvtools::utils::Timer'`

I had similar in Glslang (which I see you caught) but seems we might have to fix SPIRV-Tools as well first

--- a/.github/workflows/vvl.yml
+++ b/.github/workflows/vvl.yml
@@ -142,6 +142,28 @@ jobs:
         env:
           VK_KHRONOS_PROFILES_PROFILE_FILE: ${{ github.workspace }}/tests/device_profiles/max_profile.json
 
+  linux-ubsan:
+    needs: check_vvl
+    runs-on: ubuntu-24.04
+    name: "linux (undefined behavior sanitizer)"
+    strategy:
+      fail-fast: false
+    steps:
+      - uses: actions/checkout@v4
+      - uses: lukka/get-cmake@latest
+      - uses: hendrikmuhs/ccache-action@v1.2
+        with:
+          key: linux-ubsan
+      - run: sudo apt-get -qq update && sudo apt-get install -y libwayland-dev xorg-dev
+      - run: python scripts/tests.py --build --config debug --cmake='-DVVL_ENABLE_UBSAN=ON'
+        env:
+          CMAKE_C_COMPILER_LAUNCHER: ccache
+          CMAKE_CXX_COMPILER_LAUNCHER: ccache
+      - name: Test Max Profile
+        run: python scripts/tests.py --test
+        env:
+          VK_KHRONOS_PROFILES_PROFILE_FILE: ${{ github.workspace }}/tests/device_profiles/max_profile.json
+
   # Ensure we can build on an older Ubuntu distro with an older version of CMake.
   linux_back_compat:
     needs: check_vvl

--- a/BUILD.md
+++ b/BUILD.md
@@ -330,6 +330,10 @@ export CXXFLAGS=-fsanitize=address
 export LDFLAGS=-fsanitize=address
 ```
 
+[UBSAN (Undefined Behavior Sanitization)](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) has become a part of our CI process to ensure high quality code.
+
+For simplicity use `-D VVL_ENABLE_UBSAN=ON` to enable this sanitization method.
+
 [TSAN (Thread Sanitization)](https://clang.llvm.org/docs/ThreadSanitizer.html) has become a part of our CI process to detect data race bugs.
 
 ```bash

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -149,6 +149,18 @@ if (VVL_ENABLE_ASAN)
     endif()
 endif()
 
+option(VVL_ENABLE_UBSAN "Use undefined behavior sanitization")
+if (VVL_ENABLE_UBSAN)
+    if (NOT MSVC)
+        add_compile_options(-fsanitize=undefined)
+        add_link_options(-fsanitize=undefined)
+
+        # VVL's test-suites deliberately use invalid enum values.
+        # This creates a lot of false positives.
+        add_compile_options(-fno-sanitize=enum)
+    endif()
+endif()
+
 if(${CMAKE_CXX_COMPILER_ID} MATCHES "(GNU|Clang)")
     add_compile_options(
         -Wall

--- a/layers/core_checks/cc_descriptor.cpp
+++ b/layers/core_checks/cc_descriptor.cpp
@@ -2096,8 +2096,11 @@ bool CoreChecks::VerifyWriteUpdateContents(const vvl::DescriptorSet &dst_set, co
         }
         case VK_DESCRIPTOR_TYPE_SAMPLER: {
             auto iter = dst_set.FindDescriptor(update.dstBinding, update.dstArrayElement);
-            const vvl::SamplerDescriptor &desc = (const vvl::SamplerDescriptor &)*iter;
-            if (desc.IsImmutableSampler() && !is_push_descriptor) {
+            const vvl::SamplerDescriptor *desc = dynamic_cast<const vvl::SamplerDescriptor *>(&(*iter));
+            if (desc == nullptr) {
+                break;
+            }
+            if (desc->IsImmutableSampler() && !is_push_descriptor) {
                 skip |=
                     LogError("VUID-VkWriteDescriptorSet-descriptorType-02752", update.dstSet, write_loc.dot(Field::descriptorType),
                              "is VK_DESCRIPTOR_TYPE_SAMPLER but can't update the immutable sampler from %s.",

--- a/layers/core_checks/cc_drawdispatch.cpp
+++ b/layers/core_checks/cc_drawdispatch.cpp
@@ -273,17 +273,24 @@ bool CoreChecks::PreCallValidateCmdDrawMultiIndexedEXT(VkCommandBuffer commandBu
     skip |= ValidateActionState(cb_state, VK_PIPELINE_BIND_POINT_GRAPHICS, error_obj.location);
     skip |= ValidateVTGShaderStages(cb_state, error_obj.location);
 
+    bool invalid_stride = false;
     if (drawCount > 1) {
-        skip |= ValidateCmdDrawStrideWithStruct(cb_state, "VUID-vkCmdDrawMultiIndexedEXT-drawCount-09629", stride,
-                                                Struct::VkMultiDrawIndexedInfoEXT, sizeof(VkMultiDrawIndexedInfoEXT),
-                                                error_obj.location);
+        invalid_stride = ValidateCmdDrawStrideWithStruct(cb_state, "VUID-vkCmdDrawMultiIndexedEXT-drawCount-09629", stride,
+                                                         Struct::VkMultiDrawIndexedInfoEXT, sizeof(VkMultiDrawIndexedInfoEXT),
+                                                         error_obj.location);
     }
+    skip |= invalid_stride;
 
     // only index into pIndexInfo if we know parameters are sane
     if (drawCount != 0 && !pIndexInfo) {
         skip |= LogError("VUID-vkCmdDrawMultiIndexedEXT-drawCount-04940", cb_state.GetObjectList(VK_PIPELINE_BIND_POINT_GRAPHICS),
                          error_obj.location.dot(Field::drawCount), "is %" PRIu32 " but pIndexInfo is NULL.", drawCount);
     } else {
+        // Continuing on from this point invokes undefined behavior due to invalid stride size.
+        if (invalid_stride) {
+            return skip;
+        }
+
         const auto info_bytes = reinterpret_cast<const char *>(pIndexInfo);
         for (uint32_t i = 0; i < drawCount; i++) {
             const auto info_ptr = reinterpret_cast<const VkMultiDrawIndexedInfoEXT *>(info_bytes + i * stride);

--- a/layers/core_checks/cc_pipeline_graphics.cpp
+++ b/layers/core_checks/cc_pipeline_graphics.cpp
@@ -695,13 +695,17 @@ bool CoreChecks::ValidateGraphicsPipelineLibrary(const vvl::Pipeline &pipeline, 
             frag_shader_info.flags =
                 (pipeline.FragmentShaderPipelineLayoutState()) ? pipeline.FragmentShaderPipelineLayoutState()->CreateFlags() : 0;
             frag_shader_info.layout = pipeline.FragmentShaderPipelineLayoutState().get();
-            frag_shader_info.ms_state = pipeline.fragment_shader_state->ms_state.get()->ptr();
+            if (pipeline.fragment_shader_state->ms_state) {
+                frag_shader_info.ms_state = pipeline.fragment_shader_state->ms_state.get()->ptr();
+            }
             frag_shader_info.shading_rate_state =
                 vku::FindStructInPNextChain<VkPipelineFragmentShadingRateStateCreateInfoKHR>(pipeline_ci.pNext);
         }
         if (gpl_info->flags & VK_GRAPHICS_PIPELINE_LIBRARY_FRAGMENT_OUTPUT_INTERFACE_BIT_EXT) {
             frag_output_info.init = GPLInitType::gpl_flags;
-            frag_output_info.ms_state = pipeline.fragment_output_state->ms_state.get()->ptr();
+            if (pipeline.fragment_output_state->ms_state) {
+                frag_output_info.ms_state = pipeline.fragment_output_state->ms_state.get()->ptr();
+            }
         }
     }
 
@@ -730,13 +734,17 @@ bool CoreChecks::ValidateGraphicsPipelineLibrary(const vvl::Pipeline &pipeline, 
                     frag_shader_info.flags = layout_state->CreateFlags();
                     frag_shader_info.layout = layout_state.get();
                 }
-                frag_shader_info.ms_state = lib->fragment_shader_state->ms_state.get()->ptr();
+                if (lib->fragment_shader_state->ms_state) {
+                    frag_shader_info.ms_state = lib->fragment_shader_state->ms_state.get()->ptr();
+                }
                 frag_shader_info.shading_rate_state =
                     vku::FindStructInPNextChain<VkPipelineFragmentShadingRateStateCreateInfoKHR>(lib_ci.pNext);
             }
             if (lib->graphics_lib_type & VK_GRAPHICS_PIPELINE_LIBRARY_FRAGMENT_OUTPUT_INTERFACE_BIT_EXT) {
                 frag_output_info.init = GPLInitType::link_libraries;
-                frag_output_info.ms_state = lib->fragment_output_state->ms_state.get()->ptr();
+                if (lib->fragment_output_state->ms_state) {
+                    frag_output_info.ms_state = lib->fragment_output_state->ms_state.get()->ptr();
+                }
             }
         }
     }

--- a/layers/state_tracker/cmd_buffer_state.h
+++ b/layers/state_tracker/cmd_buffer_state.h
@@ -217,82 +217,82 @@ class CommandBuffer : public RefcountedStateObject {
     // These are values that are being set with vkCmdSet* tied to a command buffer
     struct DynamicStateValue {
         // VK_DYNAMIC_STATE_STENCIL_WRITE_MASK
-        uint32_t write_mask_front;
-        uint32_t write_mask_back;
+        uint32_t write_mask_front{};
+        uint32_t write_mask_back{};
         // VK_DYNAMIC_STATE_DEPTH_WRITE_ENABLE
-        bool depth_write_enable;
+        bool depth_write_enable{};
         // VK_DYNAMIC_STATE_DEPTH_TEST_ENABLE
-        bool depth_test_enable;
+        bool depth_test_enable{};
         // VK_DYNAMIC_STATE_DEPTH_BOUNDS_TEST_ENABLE
-        bool depth_bounds_test_enable;
+        bool depth_bounds_test_enable{};
         // VK_DYNAMIC_STATE_STENCIL_TEST_ENABLE
-        bool stencil_test_enable;
+        bool stencil_test_enable{};
         // VK_DYNAMIC_STATE_STENCIL_OP
-        VkStencilOp fail_op_front;
-        VkStencilOp pass_op_front;
-        VkStencilOp depth_fail_op_front;
-        VkStencilOp fail_op_back;
-        VkStencilOp pass_op_back;
-        VkStencilOp depth_fail_op_back;
+        VkStencilOp fail_op_front{};
+        VkStencilOp pass_op_front{};
+        VkStencilOp depth_fail_op_front{};
+        VkStencilOp fail_op_back{};
+        VkStencilOp pass_op_back{};
+        VkStencilOp depth_fail_op_back{};
         // VK_DYNAMIC_STATE_CULL_MODE
-        VkCullModeFlags cull_mode;
+        VkCullModeFlags cull_mode{};
         // VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY
-        VkPrimitiveTopology primitive_topology;
+        VkPrimitiveTopology primitive_topology{};
         // VK_DYNAMIC_STATE_SAMPLE_LOCATIONS_EXT
-        VkSampleLocationsInfoEXT sample_locations_info;
+        VkSampleLocationsInfoEXT sample_locations_info{};
         // VK_DYNAMIC_STATE_DISCARD_RECTANGLE_ENABLE_EXT
-        bool discard_rectangle_enable;
+        bool discard_rectangle_enable{};
         // VK_DYNAMIC_STATE_DISCARD_RECTANGLE_EXT
         // maxDiscardRectangles is at max 8 on all known implementations currently
-        std::bitset<32> discard_rectangles;
+        std::bitset<32> discard_rectangles{};
         // VK_DYNAMIC_STATE_POLYGON_MODE_EXT
-        VkPolygonMode polygon_mode;
+        VkPolygonMode polygon_mode{};
         // VK_DYNAMIC_STATE_RASTERIZATION_SAMPLES_EXT
-        VkSampleCountFlagBits rasterization_samples;
+        VkSampleCountFlagBits rasterization_samples{};
         // VK_DYNAMIC_STATE_RASTERIZATION_STREAM_EXT
-        uint32_t rasterization_stream;
+        uint32_t rasterization_stream{};
         // VK_DYNAMIC_STATE_SAMPLE_MASK_EXT
-        VkSampleCountFlagBits samples_mask_samples;
+        VkSampleCountFlagBits samples_mask_samples{};
         // VK_DYNAMIC_STATE_LINE_RASTERIZATION_MODE_KHR
-        VkLineRasterizationMode line_rasterization_mode;
+        VkLineRasterizationMode line_rasterization_mode{};
         // VK_DYNAMIC_STATE_LINE_STIPPLE_ENABLE_EXT
-        bool stippled_line_enable;
+        bool stippled_line_enable{};
         // VK_DYNAMIC_STATE_COVERAGE_TO_COLOR_ENABLE_NV
-        bool coverage_to_color_enable;
+        bool coverage_to_color_enable{};
         // VK_DYNAMIC_STATE_COVERAGE_TO_COLOR_LOCATION_NV
-        uint32_t coverage_to_color_location;
+        uint32_t coverage_to_color_location{};
         // VK_DYNAMIC_STATE_COVERAGE_MODULATION_MODE_NV
-        VkCoverageModulationModeNV coverage_modulation_mode;
+        VkCoverageModulationModeNV coverage_modulation_mode{};
         // VK_DYNAMIC_STATE_COVERAGE_MODULATION_TABLE_ENABLE_NV
-        bool coverage_modulation_table_enable;
+        bool coverage_modulation_table_enable{};
         // VK_DYNAMIC_STATE_SHADING_RATE_IMAGE_ENABLE_NV
-        bool shading_rate_image_enable;
+        bool shading_rate_image_enable{};
         // VK_DYNAMIC_STATE_RASTERIZER_DISCARD_ENABLE
-        bool rasterizer_discard_enable;
+        bool rasterizer_discard_enable{};
         // VK_DYNAMIC_STATE_DEPTH_BIAS_ENABLE
-        bool depth_bias_enable = false;
+        bool depth_bias_enable{};
         // VK_DYNAMIC_STATE_DEPTH_CLAMP_ENABLE_EXT
-        bool depth_clamp_enable = false;
+        bool depth_clamp_enable{};
         // VK_DYNAMIC_STATE_ALPHA_TO_COVERAGE_ENABLE_EXT
-        bool alpha_to_coverage_enable;
+        bool alpha_to_coverage_enable{};
         // VK_DYNAMIC_STATE_ALPHA_TO_ONE_ENABLE_EXT
-        bool alpha_to_one_enable;
+        bool alpha_to_one_enable{};
         // VK_DYNAMIC_STATE_LOGIC_OP_ENABLE_EXT
-        bool logic_op_enable;
+        bool logic_op_enable{};
         // VK_DYNAMIC_STATE_FRAGMENT_SHADING_RATE_KHR
-        VkExtent2D fragment_size;
+        VkExtent2D fragment_size{};
         // VK_DYNAMIC_STATE_PRIMITIVE_RESTART_ENABLE
-        bool primitive_restart_enable;
+        bool primitive_restart_enable{};
 
-        uint32_t color_write_enable_attachment_count;
+        uint32_t color_write_enable_attachment_count{};
 
         // maxColorAttachments is at max 8 on all known implementations currently
-        std::bitset<32> color_blend_enable_attachments;              // VK_DYNAMIC_STATE_COLOR_BLEND_ENABLE_EXT
-        std::bitset<32> color_blend_enabled;                         // VK_DYNAMIC_STATE_COLOR_BLEND_ENABLE_EXT
-        std::bitset<32> color_blend_equation_attachments;            // VK_DYNAMIC_STATE_COLOR_BLEND_EQUATION_EXT
-        std::bitset<32> color_write_mask_attachments;                // VK_DYNAMIC_STATE_COLOR_WRITE_MASK_EXT
-        std::bitset<32> color_blend_advanced_attachments;            // VK_DYNAMIC_STATE_COLOR_BLEND_ADVANCED_EXT
-        std::bitset<32> color_write_enabled;                         // VK_DYNAMIC_STATE_COLOR_WRITE_ENABLE_EXT
+        std::bitset<32> color_blend_enable_attachments{};            // VK_DYNAMIC_STATE_COLOR_BLEND_ENABLE_EXT
+        std::bitset<32> color_blend_enabled{};                       // VK_DYNAMIC_STATE_COLOR_BLEND_ENABLE_EXT
+        std::bitset<32> color_blend_equation_attachments{};          // VK_DYNAMIC_STATE_COLOR_BLEND_EQUATION_EXT
+        std::bitset<32> color_write_mask_attachments{};              // VK_DYNAMIC_STATE_COLOR_WRITE_MASK_EXT
+        std::bitset<32> color_blend_advanced_attachments{};          // VK_DYNAMIC_STATE_COLOR_BLEND_ADVANCED_EXT
+        std::bitset<32> color_write_enabled{};                       // VK_DYNAMIC_STATE_COLOR_WRITE_ENABLE_EXT
         std::vector<VkColorBlendEquationEXT> color_blend_equations;  // VK_DYNAMIC_STATE_COLOR_BLEND_EQUATION_EXT
         std::vector<VkColorComponentFlags> color_write_masks;        // VK_DYNAMIC_STATE_COLOR_WRITE_MASK_EXT
 
@@ -300,35 +300,35 @@ class CommandBuffer : public RefcountedStateObject {
         vvl::unordered_map<uint32_t, VertexBindingState> vertex_bindings;
 
         // VK_DYNAMIC_STATE_CONSERVATIVE_RASTERIZATION_MODE_EXT
-        VkConservativeRasterizationModeEXT conservative_rasterization_mode;
+        VkConservativeRasterizationModeEXT conservative_rasterization_mode{};
         // VK_DYNAMIC_STATE_SAMPLE_LOCATIONS_ENABLE_EXT
-        bool sample_locations_enable;
+        bool sample_locations_enable{};
         // VK_DYNAMIC_STATE_ATTACHMENT_FEEDBACK_LOOP_ENABLE_EXT
-        VkImageAspectFlags attachment_feedback_loop_enable;
+        VkImageAspectFlags attachment_feedback_loop_enable{};
 
         // VK_DYNAMIC_STATE_VIEWPORT
         std::vector<VkViewport> viewports;
         // and VK_DYNAMIC_STATE_VIEWPORT_WITH_COUNT
-        uint32_t viewport_count;
+        uint32_t viewport_count{};
         // VK_DYNAMIC_STATE_SCISSOR_WITH_COUNT
-        uint32_t scissor_count;
+        uint32_t scissor_count{};
         // VK_DYNAMIC_STATE_VIEWPORT_W_SCALING_NV
         std::vector<VkViewportWScalingNV> viewport_w_scalings;
-        uint32_t viewport_w_scaling_first;
-        uint32_t viewport_w_scaling_count;
+        uint32_t viewport_w_scaling_first{};
+        uint32_t viewport_w_scaling_count{};
         // VK_DYNAMIC_STATE_VIEWPORT_W_SCALING_ENABLE
-        bool viewport_w_scaling_enable;
+        bool viewport_w_scaling_enable{};
         // VK_DYNAMIC_STATE_VIEWPORT_SWIZZLE_NV
-        uint32_t viewport_swizzle_count;
+        uint32_t viewport_swizzle_count{};
         // VK_DYNAMIC_STATE_VIEWPORT_SHADING_RATE_PALETTE_NV
-        uint32_t shading_rate_palette_count;
+        uint32_t shading_rate_palette_count{};
         // VK_DYNAMIC_STATE_EXCLUSIVE_SCISSOR_ENABLE_NV
-        uint32_t exclusive_scissor_enable_first;
-        uint32_t exclusive_scissor_enable_count;
-        std::vector<VkBool32> exclusive_scissor_enables;
+        uint32_t exclusive_scissor_enable_first{};
+        uint32_t exclusive_scissor_enable_count{};
+        std::vector<VkBool32> exclusive_scissor_enables{};
         // VK_DYNAMIC_STATE_EXCLUSIVE_SCISSOR_NV
-        uint32_t exclusive_scissor_first;
-        uint32_t exclusive_scissor_count;
+        uint32_t exclusive_scissor_first{};
+        uint32_t exclusive_scissor_count{};
         std::vector<VkRect2D> exclusive_scissors;
 
         // When the Command Buffer resets, the value most things in this struct don't matter because if they are read without

--- a/layers/state_tracker/pipeline_state.cpp
+++ b/layers/state_tracker/pipeline_state.cpp
@@ -453,6 +453,10 @@ static bool IgnoreColorAttachments(const Device &state_data, Pipeline &pipe_stat
 }
 
 static bool UsesShaderModuleId(const Pipeline &pipe_state) {
+    if (pipe_state.shader_stages_ci.data() == nullptr) {
+        return false;
+    }
+
     for (const auto &stage_ci : pipe_state.shader_stages_ci) {
         // if using GPL, can have null pStages
         if (stage_ci.ptr()) {

--- a/layers/stateless/sl_descriptor.cpp
+++ b/layers/stateless/sl_descriptor.cpp
@@ -86,7 +86,7 @@ bool Device::ValidateCoarseSampleOrderCustomNV(const VkCoarseSampleOrderCustomNV
     // guarantee that 64 bits is enough, but practically it's unlikely for an
     // implementation to support more than 32 bits for samplemask.
     assert(phys_dev_ext_props.shading_rate_image_props.shadingRateMaxCoarseSamples <= 64);
-    uint64_t sample_locations_mask = 0;
+    std::bitset<64> sample_locations_mask = 0;
     for (uint32_t i = 0; i < order.sampleLocationCount; ++i) {
         const VkCoarseSampleLocationNV *sample_loc = &order.pSampleLocations[i];
         if (sample_loc->pixelX >= sample_order_info->width) {
@@ -108,10 +108,13 @@ bool Device::ValidateCoarseSampleOrderCustomNV(const VkCoarseSampleOrderCustomNV
         }
         uint32_t idx =
             sample_loc->sample + order.sampleCount * (sample_loc->pixelX + sample_order_info->width * sample_loc->pixelY);
-        sample_locations_mask |= 1ULL << idx;
+        // Account for idx being greater than or equal to 64 to prevent undefined behavior
+        if (idx < sample_locations_mask.size()) {
+            sample_locations_mask[idx] = 1;
+        }
     }
 
-    uint64_t expected_mask = (order.sampleLocationCount == 64) ? ~0ULL : ((1ULL << order.sampleLocationCount) - 1);
+    std::bitset<64> expected_mask = (order.sampleLocationCount == 64) ? ~0ULL : ((1ULL << order.sampleLocationCount) - 1);
     if (sample_locations_mask != expected_mask) {
         skip |= LogError(
             "VUID-VkCoarseSampleOrderCustomNV-pSampleLocations-02077", device, order_loc,

--- a/scripts/CMakeLists.txt
+++ b/scripts/CMakeLists.txt
@@ -58,6 +58,10 @@ if (UPDATE_DEPS)
         list(APPEND update_dep_command "--asan")
     endif()
 
+    if (VVL_ENABLE_UBSAN)
+        list(APPEND update_dep_command "--ubsan")
+    endif()
+
     if (NOT BUILD_TESTS)
         list(APPEND update_dep_command "--optional=tests")
     endif()

--- a/scripts/tests.py
+++ b/scripts/tests.py
@@ -176,6 +176,9 @@ def RunVVLTests(args):
 
     lvt_env = dict(os.environ)
 
+    # Needed for undefined behavior sanitizor to actually cause program to fail.
+    lvt_env['UBSAN_OPTIONS'] = 'abort_on_error=1:halt_on_error=1'
+
     # Because we installed everything to CI_INSTALL_DIR all the libraries/json files are in pre-determined locations
     # defined by GNUInstallDirs. This makes setting VK_LAYER_PATH and other environment variables trivial/robust.
     if common_ci.IsWindows():

--- a/scripts/update_deps.py
+++ b/scripts/update_deps.py
@@ -510,6 +510,13 @@ class GoodRepo(object):
             if platform.system() != 'Windows':
                 os.environ['LDFLAGS'] = '-fsanitize=address'
 
+        # Optionally build dependencies with UBSAN enabled
+        if self._args.ubsan:
+            cmake_cmd.append('-D CMAKE_CXX_FLAGS=-fsanitize=undefined -fno-sanitize=enum')
+            cmake_cmd.append('-D CMAKE_C_FLAGS=-fsanitize=undefined')
+            cmake_cmd.append('-D ENABLE_RTTI=ON')
+            os.environ['LDFLAGS'] = '-fsanitize=undefined'
+
         # Use the CMake -A option to select the platform architecture
         # without needing a Visual Studio generator.
         if platform.system() == 'Windows' and self._args.generator != "Ninja":
@@ -747,6 +754,12 @@ def main():
         dest='asan',
         action='store_true',
         help="Build dependencies with ASAN enabled",
+        default=False)
+    parser.add_argument(
+        '--ubsan',
+        dest='ubsan',
+        action='store_true',
+        help="Build dependencies with UBSAN enabled",
         default=False)
 
     args = parser.parse_args()

--- a/tests/framework/video_objects.h
+++ b/tests/framework/video_objects.h
@@ -1403,7 +1403,9 @@ class VideoDecodeInfo : public VideoOpParams<VkVideoDecodeInfoKHR> {
 
                 codec_info_.decode_av1.std_picture_info = {};
                 // We will simply use the DPB slot index as the VBI slot index for the purposes of testing
-                codec_info_.decode_av1.std_picture_info.refresh_frame_flags = static_cast<uint8_t>(1 << slot_index);
+                if (slot_index >= 0) {
+                    codec_info_.decode_av1.std_picture_info.refresh_frame_flags = static_cast<uint8_t>(1 << slot_index);
+                }
 
                 reconstructed_.pNext = &codec_info_.decode_av1.setup_slot_info;
 

--- a/tests/unit/command.cpp
+++ b/tests/unit/command.cpp
@@ -1678,7 +1678,6 @@ TEST_F(NegativeCommand, MultiDraw) {
     m_errorMonitor->SetDesiredError("VUID-vkCmdDrawMultiEXT-drawCount-09628");
     vk::CmdDrawMultiEXT(m_command_buffer.handle(), 3, multi_draws, 1, 0, sizeof(VkMultiDrawInfoEXT) + 1);
     m_errorMonitor->VerifyFound();
-    m_errorMonitor->SetDesiredError("VUID-vkCmdDrawMultiIndexedEXT-robustBufferAccess2-08798", 2);
     m_errorMonitor->SetDesiredError("VUID-vkCmdDrawMultiIndexedEXT-drawCount-09629");
     vk::CmdDrawMultiIndexedEXT(m_command_buffer.handle(), 3, multi_draw_indices, 1, 0, sizeof(VkMultiDrawIndexedInfoEXT) + 1, 0);
     m_errorMonitor->VerifyFound();

